commercial/app/model/merchandise/books/BookFinder.scala

```diff
@@ -2,67 +2,56 @@ package commercial.model.merchandise.books
 
 import akka.actor.ActorSystem
 import akka.pattern.CircuitBreaker
+import akka.util.Timeout
 import commercial.model.feeds.{FeedParseException, FeedReadException, FeedReader, FeedRequest}
-import common.Logging
+import commercial.model.merchandise.Book
+import common.{AkkaAgent, Logging}
 import conf.Configuration
 import conf.switches.Switches.BookLookupSwitch
-import commercial.model.merchandise.Book
 import play.api.libs.json._
 import play.api.libs.oauth.{ConsumerKey, OAuthCalculator, RequestToken}
 import play.api.libs.ws.{WSClient, WSSignatureCalculator}
-import shade.memcached.{Memcached, MemcachedCodecs, Configuration => MemcachedConfiguration}
 
 import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.control.NonFatal
 
 class BookFinder(actorSystem: ActorSystem, magentoService: MagentoService) extends Logging {
 
-  private implicit lazy val executionContext = actorSystem.dispatchers.lookup("akka.actor.memcached")
-
-  private lazy val defaultBookDataCache = new MemcachedBookDataCache(actorSystem)
-
-  def findByIsbn(isbn: String,
-                 cache: BookDataCache = defaultBookDataCache,
-                 lookup: String => Future[Option[JsValue]] = magentoService.findByIsbn):
-  Future[Option[Book]] = {
-
-    def cachedBook(bookData: JsValue): Future[Option[Book]] = {
-      Future.successful {
-        bookData match {
-          case JsNull => None
-          case _ =>
-            bookData.validate[Book] fold(
-              invalid => {
-                log.error(Json.stringify(JsError.toJson(invalid)))
-                None
-              },
-              book => Some(book))
-        }
+  private implicit val bookActorExecutionContext: ExecutionContext = actorSystem.dispatchers.lookup("akka.actor.book-lookup")
+  private implicit val bookActorTimeout: Timeout = 0.2.seconds
+  private implicit val magentoServiceImplicit = magentoService
+
+  def findByIsbn(isbn: String): Option[Book] = BookAgent.get(isbn) map { _.as[Book] }
+}
+
+object BookAgent extends Logging {
+
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
+  private lazy val lookupJobs = AkkaAgent(Set.empty[String])
```

Comment:
It is to prevent the Magento Lookup of the same ISBN and is used [here](https://github.com/guardian/frontend/pull/16533/files/482361cfc65400b08035d4016d71c4c3ed740f64#diff-b8caa4a82242d9b3eece8ab4bdf98828R42).

Scenario: we get request `A` in for ISBN `123`. We don't have it in the cache so we use Magento to look it up. Whilst the Magento service is doing this (can take seconds), request `B` comes in from another user for the same ISBN `123`. We needn't go to Magento again as we know that it is already being retrieved.